cpp/src/arrow/util/decimal.cc

```diff
@@ -156,7 +162,9 @@ struct DecimalRealConversion : public BaseDecimalRealConversion {
         // NOTE: if `precision` is the full precision then the algorithm will
         // lose the last digit. If `precision` is almost the full precision,
         // there can be an off-by-one error due to rounding.
-        const int mul_step = std::max(1, kMaxPrecision - precision);
+        constexpr int is_dec32_or_dec64 =
+            DecimalType::kByteWidth <= BasicDecimal64::kByteWidth;
+        const int mul_step = std::max(1, kMaxPrecision - precision - is_dec32_or_dec64);
```

Comment:
With decimal32 and decimal64 I needed to reduce the `mul_step` by 1 in order to eliminate off-by-one rounding errors. I didn't want to unconditionally add extra operations to decimal128/256 by lowering their `mul_step` so I only did it for decimal32/64

Reply:
That's several things that need to be arbitrarily lowered, and sounds a bit unexpected.
Random hacks like this are a bad smell, especially if there's no detailed explanation other that "it works better".

Reply:
That said, if you don't want to debug these now, I would favor removing the hacks, adding temporary workarounds in the tests, and opening an issue together with value(s) reproducing the problem. Then this can be investigated (by you, me or @bkietz for example :-)).

Reply:
I've adjusted the tests to simply not hit this case anymore with the off-by-one rounding issue for decimal64. 

Let me know if you prefer that to the above hack with explanation.